### PR TITLE
feat(aws-autoscaling): allow setting spotPrice

### LINKED ISSUE #2208: Support Spot instances
ECS currently supports Spot instance fleets -- to support this feature, we would need Spot instance L2s in ec2 construct lib.

### PR DESCRIPTION
**Pull request by SoManyHs**
Fixes #2208.

----

### Pull Request Checklist

* [X] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [X] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [X] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [X] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

**Comment by SoManyHs** [comment]
Note: Is it worth refactoring `instanceType` to be in the `CommonAutoScalingGroupProps`, since it's used in the three places that extend that interface? (EKS cluster, ECS cluster, and regular ASG).

**Comment by pkandasamy91** [comment]
> Note: Is it worth refactoring `instanceType` to be in the `CommonAutoScalingGroupProps`, since it's used in the three places that extend that interface? (EKS cluster, ECS cluster, and regular ASG).

I think it makes sense to refactor into `CommonAutoScalingGroupProps`. Currently, every properties object that extends `CommonAutoScalingGroupProps` has `instanceType` as a property.

**Comment by SoManyHs** [comment]
From talking offline with @rix0rrr:
The common properties are all optional on purposes, so you can always safely extend them to add more customization without restricting the rest of your api.

since `instanceType` is a required property, it wouldn't make sense to add them to the common props on that basis. However, perhaps another extendable interface with required properties would be useful and maybe can be done in a future PR.

**Comment by pkandasamy91** [comment]
Makes sense, thanks!

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 1]
```diff
@@ -174,6 +174,14 @@ export interface AutoScalingGroupProps extends CommonAutoScalingGroupProps {
...
    */
   readonly role?: iam.IRole;
+
+  /**
+   * The maximum hourly price to be paid for any Spot Instance launched to fulfill the request. Spot Instances are
+   * launched when the price you specify exceeds the current Spot market price.
+   *
+   * @default none
+   */
+  readonly spotPrice?: string;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Wouldn't it have made sense to add this to `CommonAutoScalingGroupProps`?

Also, is this literally all you need to do to support spot instance fleets? What happens if the price exceeds the number you put here? Will there be no instances running? How does that interact with MinSize?

I would appreciate a couple more lines on the behavior here, or a link to the documentation.

Comment (rix0rrr) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Also, in the interest of annotating numbers with their unit, I feel this should be called `spotPriceUsd` or something, no?

Comment (SoManyHs) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Yes, I added an integ test to show that this in fact works.

Adding USD feels a little verbose, but if we do that elsewhere I can change it. And yes, can also move this to `CommonAutoScalingGroupProps`.

**Pull request merged**